src/newusers.c

```diff
@@ -1280,17 +1267,19 @@ int main (int argc, char **argv)
 		if (is_sub_uid && want_subuids() && !local_sub_uid_assigned(fields[0])) {
 			uid_t sub_uid_start = 0;
 			unsigned long sub_uid_count = 0;
-			if (find_new_sub_uids(&sub_uid_start, &sub_uid_count) == 0) {
-				if (sub_uid_add(fields[0], sub_uid_start, sub_uid_count) == 0) {
-					fprintf (stderr,
-						_("%s: failed to prepare new %s entry\n"),
-						Prog, sub_uid_dbname ());
-				}
-			} else {
+			if (find_new_sub_uids(&sub_uid_start, &sub_uid_count) != 0)
+			{
 				fprintf (stderr,
 					_("%s: can't find subordinate user range\n"),
 					Prog);
-				errors++;
+				fail_exit (EXIT_FAILURE);
+			}
+			if (sub_uid_add(fields[0], sub_uid_start, sub_uid_count) == 0)
```

Comment:
Am I upside down, or should this (and the sub_gid_add check below) be erroring on != 0 ?

Reply:
Considering precedents on this patch set, I'd bet on me being upside down.

Let's see the transformation:

```c
if (find() == 0) {
    if (sub() == 0) {
        fprintf("failed");
    }
} else {
    fprintf("can't");
    errors++;
}
```

```c
if (find() != 0) {
    fprintf("can't");
    errors++;
} else {
    if (sub() == 0) {
        fprintf("failed");
    }
}
```

```c
if (find() != 0) {
    fprintf("can't");
    fail();
} else {
    if (sub() == 0) {
        fprintf("failed");
        fail();
    }
}
```

```c
if (find() != 0) {
    fprintf("can't");
    fail();
}
if (sub() == 0) {
    fprintf("failed");
    fail();
}
```

Which seems to coincide with the patch.  It seems I wasn't wrong this time (unless I missed something) :)

Reply:
Jiminy cricket, you are right.  I was only comparing the previous code with the new through the patch.  But the previous code was wrong.

The error values of functions in this project are abysmal.  I should get back to working on the rust version where we can use tuples to avoid that  :)

So yes, with sub_uid_add(), 1 is success, failure can be 0 with errno set OR -errno.

Yuck.

Reply:
Oh, not even.  I still also misread the original code.

Reply:
Ok, so if get_subid_nss_handle() fails, this will be wrong, since you are checking for == 0 not <= 0.

I think the better thing is to fix the (my, I'm pretty sure) sub_gid_add() bug where it returns -ERANGE instead of setting errno to ERANGE and returning 0.

Reply:
Do you want to send me a patch so I include it in the PR?  Or maybe you prefer a separate PR after this one and the alloc one have been merged?

And yes, I've seen a lot of weirdness in this project in return codes.  But please don't despair, and go to Rust so soon.  :p

Cheers!

Reply:
Please do include it in this pr, so i can merge it all at once.

Thanks!

Reply:
Oh, I meant that you send me a patch (e.g., via email; or maybe suggest it in GH, if you know how it works --I don't--) so I include it in the PR.  I don't know how to fix it exactly :-)

Reply:
Oh, sorry, I see :)

Well - let's just merge this as is, and I'll file a PR tomorrow for that.